### PR TITLE
Add more settings to yaml file

### PR DESCRIPTION
**Pull request by pieper126**
All settings should reside in either the configuration file or in parameter.

### CODE REVIEW

File: main.py
Comment (serroba) [inline main.py:1]:
```class Client:
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port

    def connect(self):
        raise NotImplementedError


class ModbusClient(Client):
    def __init__(self, configuration: Configuration, args):
        super().__init__(args.modbus_host, args.modbus_port)

    def connect(self):
        # Implementation of Modbus connection
        pass


class MqttClient(Client):
    def __init__(self, args):
        super().__init__(args.mqtt_host, args.mqtt_port)

    def connect(self):
        # Implementation of MQTT connection
        pass


def handle_args(configuration: Configuration):
    # Create the argument parser
    parser = argparse.ArgumentParser(description='remote commands handler')

    # Add the optional arguments
    parser.add_argument('--configuration_path', required=True, help='Configuration path')
    parser.add_argument('--modbus_port', default=configuration.modbus_settings.port, type=int, help='Modbus port')
    parser.add_argument('--modbus_host', default=configuration.modbus_settings.host, help='Modbus host')
    parser.add_argument('--mqtt_port', default=configuration.mqtt_settings.port, type=int, help='MQTT port')
    parser.add_argument('--mqtt_host', default=configuration.mqtt_settings.host, help='MQTT host')

    return parser.parse_args()


def main():
    configuration = get_configuration()  # A function to load configuration from the YAML file
    args = handle_args(configuration)

    modbus_client = ModbusClient(configuration, args)
    mqtt_client = MqttClient(args)

    # Now you can connect your clients or whatever you want to do
```

Comment (serroba) [reply main.py:1]: Have a look at how you can avoid the `if/else` by using the `default` parameter as part of the arguments parser.

File: config/example_configuration.yml
Comment (serroba) [inline config/example_configuration.yml:1]: If this is the main config file, then it shouldn't be named `example_`.
Here, you should be providing just a config with common default values. In addition to that, we should be adding more careful documentation, but we can do that over time.

File: main.py [hunk 1]
```diff
@@ -1,25 +1,69 @@
...
-MODBUS_HOST = os.getenv('MODBUS_HOST')
-MODBUS_PORT = int(os.getenv('MODBUS_PORT'))
-COMMANDS_TOPIC = os.getenv('COMMANDS_TOPIC')
-CONFIGURATION_PATH = os.getenv('CONFIGURATION_PATH', "config/example_configuration.yaml")
+def handle_args():
+    # Create the argument parser
+    parser = argparse.ArgumentParser(description='remote commands handler')
+
+    # Add the optional arguments
+    parser.add_argument('--configuration_path', required=True, help='Configuration path')
```
Comment (serroba) [inline main.py:16]: This one should have a default value pointing to the `config/configuration.yaml` file.

Comment (serroba) [reply main.py:16]:
```suggestion
    parser.add_argument('--configuration_path', required=True, help='Configuration path', default='config/configuration.yaml')
```

I guess

File: app/modbus_client.py [hunk 2]
```diff
@@ -1,25 +1,26 @@
...
     def __init__(self, configuration: Configuration, port: int, host: str) -> None:
-        self.configuration = configuration 
+        self.configuration = configuration
         self.client = ModbusTcpClient(host, port=port)
 
     def write_coils(self, name: str, value: list[bool]):
         coil_configuration = self.configuration.get_coil(name)
         self.client.connect()
-        self.client.write_coils(coil_configuration.address, value)
+        self.client.write_coils(coil_configuration.address[0], value)
```
Comment (serroba) [inline app/modbus_client.py:13]: We are currently manually selecting the first address.
We will need to learn how we handle this when we have multiple addresses, maybe there is a native way to handle this, or maybe we need to manually write to multiple addresses.

Telegraf supports multiple address in its configuration and I think is sensible following the same pattern.

For now though, we are just writing to the first address in the document.

**Review by serroba — APPROVED** [review]

File: app/modbus_client.py (hunk 2 above)
Comment (march83) [reply app/modbus_client.py:13]: pymodbus write_coils here: https://pymodbus.readthedocs.io/en/latest/source/library/client.html#pymodbus.client.mixin.ModbusClientMixin.write_coils

...and the specific implementation here: https://pymodbus.readthedocs.io/en/latest/source/library/pymodbus.html#pymodbus.bit_write_message.WriteMultipleCoilsRequest

This write_coil**s** function is useful only if there is a need to write a sequence of bools to a single address which I don't currently see as likely.

More likely is a sequence of payloads to different addresses - in that case the ability to send all commands (possibly coils or registers) all within the same connect -> close session might be useful?

Comment (serroba) [reply app/modbus_client.py:13]: I am not sure this is correct @march83 

```
    def write_coils(
        self,
        address: int,
        values: Union[List[bool], bool],
        slave: int = 0,
        **kwargs: Any,
    ) -> ModbusResponse:
        """Write coils (code 0x0F).

        :param address: Start address to write to
        :param values: List of booleans to write, or a single boolean to write
        :param slave: (optional) Modbus slave ID
        :param kwargs: (optional) Experimental parameters.
        :raises ModbusException:
        """
        return self.execute(
            pdu_bit_write.WriteMultipleCoilsRequest(address, values, slave, **kwargs)
        )
```

As you can see, the method is an step of indirection.

If you pay attention to the method signature there also:

```
class WriteMultipleCoilsRequest(ModbusRequest):
    """This function code is used to forcea sequence of coils.

    To either ON or OFF in a remote device. The Request PDU specifies the coil
    references to be forced. Coils are addressed starting at zero. Therefore
    coil numbered 1 is addressed as 0.

    The requested ON/OFF states are specified by contents of the request
    data field. A logical "1" in a bit position of the field requests the
    corresponding output to be ON. A logical "0" requests it to be OFF."
    """

    function_code = 15
    function_code_name = "write_coils"
    _rtu_byte_count_pos = 6

    def __init__(self, address=None, values=None, slave=None, **kwargs):
        """Initialize a new instance.

        :param address: The starting request address
        :param values: The values to write
        """
        ModbusRequest.__init__(self, slave=slave, **kwargs)
        self.address = address
        if not values:
```

I think if anything, we are doing the right thing for now.

Comment (serroba) [reply app/modbus_client.py:13]: image

This are all the write methods published by the `client`. 

We probably may need to switch, from using the plain `client` abstraction, to build our own if this doesn't satisfy our needs, but I would recommend to do that as a follow up PR

Comment (march83) [reply app/modbus_client.py:13]: Yep, you are right. It's a list of coils with a start address.

This aligns with modbus function code 15, makes sense now.

Comment (march83) [reply app/modbus_client.py:13]: To your original question, if there is a list of non-consecutive coils to write to then they need to be handled independently with a new function call each time.

**Pull request merged**